src/kernel/arch/x86/syscalls.zig

```diff
@@ -0,0 +1,70 @@
+const arch = @import("arch.zig");
+const testing = @import("std").testing;
+const assert = @import("std").debug.assert;
+const isr = @import("isr.zig");
+const log = @import("../../log.zig");
+
+pub const INTERRUPT: u16 = 0x80;
+pub const NUM_HANDLERS = 256;
+
+/// A syscall handler
+pub const SyscallHandler = fn (ctx: *arch.InterruptContext) void;
+/// Errors that syscall utility functions can throw
+pub const SyscallError = error {
+    SyscallExists
+};
+
+var handlers = []?SyscallHandler{null} ** NUM_HANDLERS;
+
+fn handle(ctx: *arch.InterruptContext) void {
+    // The syscall number is put in eax
+    const syscall = ctx.eax;
+    if (syscall < NUM_HANDLERS) {
+        if (handlers[syscall]) |handler| {
+            handler(ctx);
+        } else {
+            log.logWarning("Syscall {} triggered but not registered\n", syscall);
+        }
+    }
+}
+
+///
+/// Register a syscall so it  can be called by triggering interrupt 128
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall number to register the handler with
+///     IN handler: SyscallHandler - The handler to register the syscall with
+///
+pub fn registerSyscall(syscall: u8, handler: SyscallHandler) !void {
+    if (handlers[syscall]) |_|
+        return SyscallError.SyscallExists;
+    handlers[syscall] = handler;
+}
+
+///
+/// Trigger a syscall
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall to trigger. Is put in eax and must have been registered.
+///
+pub fn doSyscall(syscall: u8) void {
+    asm volatile ("mov %%ecx, %%eax" :: [syscall] "{ecx}" (syscall));
+    asm volatile ("int $0x80");
+}
+
+pub fn init() void {
+    log.logInfo("Init syscalls\n");
+    isr.registerIsr(INTERRUPT, handle);
+    log.logInfo("Done\n");
+}
+
+fn testHandler(ctx: *arch.InterruptContext) void {
+}
+
+test "registerSyscall returns SyscallExists" {
+    registerSyscall(123, testHandler) catch unreachable;
+    registerSyscall(123, testHandler) catch |err| {
+        return;
+    };
+    assert(false);
+}
```

Comment:
Could add tests for calling a syscall with ur testHandler

Reply:
I'll be able to do so once I figure out how to pass options from build.zig to the kernel to differentiate between a test build and a notmal build, as these tests would have to be done at runtime (since they depend on the CPU doing interrupts) and we don't want to be doing them for a normal build.

Reply:
Opening an issue for that would be good actually, so we don't forget